core/commonMain/src/kotlinx/serialization/PolymorphicSerializer.kt

```diff
@@ -67,14 +67,28 @@ import kotlin.reflect.*
  */
 @OptIn(ExperimentalSerializationApi::class)
 public class PolymorphicSerializer<T : Any>(override val baseClass: KClass<T>) : AbstractPolymorphicSerializer<T>() {
-    public override val descriptor: SerialDescriptor =
+
+    @PublishedApi // should we allow user access to this constructor?
+    internal constructor(
+        baseClass: KClass<T>,
+        classAnnotations: Array<Annotation>
+    ) : this(baseClass) {
+        _descriptor.annotations = classAnnotations.asList()
+    }
+
+    private val _descriptor: SerialDescriptorImpl =
```

Comment:
Suggest change here to avoid mutable descriptor:

```
public class PolymorphicSerializer<T : Any>(override val baseClass: KClass<T>) : AbstractPolymorphicSerializer<T>() {

    @PublishedApi // should we allow user access to this constructor?
    internal constructor(
        baseClass: KClass<T>,
        classAnnotations: Array<Annotation>
    ) : this(baseClass) {
        annotations = classAnnotations
    }


    private var annotations = emptyArray<Annotation>()

    public override val descriptor: SerialDescriptor =  buildSerialDescriptor("kotlinx.serialization.Polymorphic", PolymorphicKind.OPEN) {
        element("type", String.serializer().descriptor)
        element(
            "value",
            buildSerialDescriptor(
                "kotlinx.serialization.Polymorphic<${baseClass.simpleName}>",
                SerialKind.CONTEXTUAL
            )
        )
        annotations = this@PolymorphicSerializer.annotations.asList()
    }.withContext(baseClass)
```

Reply:
the problem is, `buildSerialDescriptor`, as initializing expression, is executed before secondary constructor

Reply:
See also https://github.com/Kotlin/kotlinx.serialization/pull/1515#discussion_r642580350

Reply:
Let's make it lazy then?

Reply:
Won't expect-actual help us here?

Reply:
I can try